lib/Semantic_check.ml

```diff
@@ -1,2163 +0,0 @@
-(** Semantic validation of AST*)
-
-(* Idea: check many of things related to identifiers that are hard to check
-during parsing and are in fact irrelevant for building up the parse tree *)
-
-(* To check:
-Variable in scope if used 
-Variable assigned to before dereferenced 
-Target+= can only be used in model and functions with right suffix (same for tilde etc)
-Data and parameters never assigned to
-Transformed data only assigned in td block
-Tp only assigned in Tp block
-Parameters and transformed parameters are not int(array) (silly restriction if you ask me)
-Model top level variables only assigned and read in model 
-Function arguments match signature 
-Assignments are type consistent
-Indexes have int type 
-Expressions take valid number of indices (based on their matrix/array dimensions)
-Funaps are actually functions
-For, while, for each, if constructs take expressions of valid type
-No shadowing!
-Check variable not already declared during declaration
-Check that function args and loop identifiers are not modified in function. (passed by const ref) 
-No returns outside of function definitions?
-Rng functions cannot be used in Tp or Model and only in funciton defs with the right suffix
-Print/reject expressions cannot be of type void.
-Break and continue only occur in loops.
-Array expressions must be of uniform type. (Or mix of int and real)
-Typing of ~ and target +=
-Also check whether function arguments meet data requirement.
-Maybe should also infer bounds for every indexing that happens so we know what code to generate for bound checking?
-Every trace through function body contains return statement of right type
-In case of void function, no return statements anywhere
-All function arguments are distinct
-Function applications are returning functions
-NRFunction applications are non-returning functions
-Primitives cannot be printed
-Functions cannot be printed
-User defined functions cannot be overloaded
-Function ending in _lp only where target is available
-TODO: Test that user defined functions with probability suffixes have right type.
-(Mutual) recursive functions have a definition
-Make sure return types of statements involving continue and break are correct.
-Make sure data only arguments to functions are checked properly.
-Sizes should be of level at most data.
-*)
-
-(* TODO: check that variables are assigned to before they are used! *)
-
-open Symbol_table
-open Ast
-open Stan_math_signatures
-open Operators
-open Errors
-open Type_conversion
-open Pretty_printing
```

Comment:
We had some discussion and think it might make sense to try opening Core_kernel in most places to essentially change to that standard library in a consistent fashion.

Reply:
I've been trying to switch to this, but it's been giving me more trouble than expected. I'm getting a lot of clashes. I'm starting to wonder whether it's worth it. Will try for a bit longer to give it a chance.